openforcefield/utils/toolkits.py

```diff
@@ -535,12 +583,55 @@ def _read_oemolistream_molecules(cls, oemolistream, allow_undefined_stereo, file
             oechem.OEPerceiveChiral(oemol)
             oechem.OEAssignAromaticFlags(oemol, oechem.OEAroModel_MDL)
             oechem.OE3DToInternalStereo(oemol)
-            mol = cls.from_openeye(
-                oemol,
-                allow_undefined_stereo=allow_undefined_stereo)
-            mols.append(mol)
 
-            # Check if this file may be using GAFF atom types.
+            # If this is either a multi-conformer or multi-molecule SD file, check to see if there are partial charges
+            if (oemolistream.GetFormat() == oechem.OEFormat_SDF) and hasattr(oemol, 'GetConfs'):
+                # The openFF toolkit treats each conformer in a "multiconformer" SDF as
+                # a separate molecule.
+                # https://github.com/openforcefield/openforcefield/issues/202
+                # Note that there is ambiguity about how SD data and "multiconformer" SD files should be stored.
+                # As a result, we have to do some weird stuff below, as discussed in
+                # https://docs.eyesopen.com/toolkits/python/oechemtk/oemol.html#dude-where-s-my-sd-data
+
+                # Jeff: I was unable to find a way to distinguish whether a SDF was multiconformer or not.
+                # The logic below should handle either single- or multi-conformer SDFs.
+                for conf in oemol.GetConfIter():
+                    # First, we turn "conf" into an OEMCMol (OE multiconformer mol), since OTHER file formats
+                    # really are multiconformer, and we will eventually feed this into the `from_openeye` function,
+                    # which is made to ingest multiconformer mols.
+                    this_conf_oemcmol = conf.GetMCMol()
+
+                    # Then, we take any SD data pairs that were on the oemol, and copy them on to "this_conf_oemcmol".
+                    # These SD pairs will be populated if we're dealing with a single-conformer SDF.
+                    for dp in oechem.OEGetSDDataPairs(oemol):
```

Comment:
Just to make sure I follow this bit we are saying if it is a single conformer SDF the tags are in the `oemol` and will be transferred in the first loop else if its a multi conformer SDF the first loop will have no tags as they are instead in the `conf` molecule?

Reply:
Exactly. It's super confusing, but this logic is intended to handle both scenarios.